testing/hooks.ts

```diff
@@ -0,0 +1,63 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+type THookFn = () => void | Promise<void>;
+type THook = (fn: THookFn) => void | Promise<void>;
+type HookType = "beforeEach" | "afterEach";
+
+const hooks: Partial<Record<HookType, THookFn>> = {};
+
+/**
+ * Register new testing hooks.
+ *
+ * @param {Function} fn
+ * @param {HookType} hookType
+ */
+function _addHook(fn: THookFn, hookType: HookType): void {
+  if (typeof fn !== "function") {
+    throw new TypeError(
+      "Invalid first argument. It must be a callback function.",
+    );
+  }
+  if (typeof hookType !== "string") {
+    throw new TypeError(
+      `Invalid second argument, ${hookType}. It must be a string`,
+    );
+  }
+  hooks[hookType] = fn;
+}
+
+/**
+ * Helper function to execute registred hooks.
+ */
+export const withHooks = (fn: THookFn) =>
+  async () => {
+    if (typeof fn !== "function") {
```

Comment:
Type check before creating the promise.

Reply:
why we will check before creating a promise ?